admin/class-primary-term-admin.php

```diff
@@ -0,0 +1,195 @@
+<?php
+/**
+ * @package WPSEO\Admin
+ */
+
+/**
+ * Adds the UI to change the primary term for a post
+ */
+class WPSEO_Primary_Term_Admin {
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		add_action( 'admin_footer', array( $this, 'wp_footer' ), 10 );
+
+		add_action( 'admin_enqueue_scripts', array( $this, 'enqueue_assets' ) );
+
+		add_action( 'save_post', array( $this, 'save_primary_terms' ) );
+		add_filter( 'post_link_category', array( $this, 'post_link_category' ) );
+	}
+
+	/**
+	 * Add primary term templates
+	 */
+	public function wp_footer() {
+		$taxonomies = $this->get_primary_term_taxonomies();
+
+		if ( ! empty( $taxonomies ) ) {
+			include_once WPSEO_PATH . '/admin/views/js-templates-primary-term.php';
+		}
+	}
+
+	/**
+	 * Enqueues all the assets needed for the primary term interface
+	 *
+	 * @return void
+	 */
+	public function enqueue_assets() {
+		$taxonomies = $this->get_primary_term_taxonomies();
+
+		// Only enqueue if there are taxonomies that need a primary term.
+		if ( empty( $taxonomies ) ) {
+			return;
+		}
+
+		wp_enqueue_style( 'wpseo-primary-category', plugins_url( 'css/metabox-primary-category' . WPSEO_CSSJS_SUFFIX . '.css', WPSEO_FILE ), array(), WPSEO_VERSION );
+
+		wp_enqueue_script( 'wpseo-primary-category', plugins_url( 'js/wp-seo-metabox-category' . WPSEO_CSSJS_SUFFIX . '.js', WPSEO_FILE ), array( 'jquery' ), WPSEO_VERSION, true );
+
+		$taxonomies = array_map( array( $this, 'map_taxonomies_for_js' ), $taxonomies );
+
+		$data = array(
+			'taxonomies' => $taxonomies,
+		);
+		wp_localize_script( 'wpseo-primary-category', 'wpseoPrimaryCategoryL10n', $data );
+	}
+
+	/**
+	 * Saves all selected primary terms
+	 *
+	 * @param int $post_ID Post ID to save primary terms for.
+	 */
+	public function save_primary_terms( $post_ID ) {
+		$taxonomies = $this->get_primary_term_taxonomies( $post_ID );
+
+		foreach ( $taxonomies as $taxonomy ) {
+			$this->save_primary_term( $post_ID, $taxonomy );
+		}
+	}
+
+	/**
+	 * Filters post_link_category to change the category to the chosen category by the user
+	 *
+	 * @param stdClass $category The category that is now used for the post link.
+	 *
+	 * @return array|null|object|WP_Error The category we want to use for the post link.
+	 */
+	public function post_link_category( $category ) {
+		$primary_category = new WPSEO_Primary_Term( 'category', get_the_ID() );
+		$primary_category = $primary_category->get_primary_term();
+
+		if ( false !== $primary_category && $primary_category !== $category->cat_ID ) {
+			$category = get_category( $primary_category );
+		}
+
+		return $category;
+	}
+
+	/**
+	 * Returns an array suitable for use in the javascript
+	 *
+	 * @param stdClass $taxonomy The taxonomy to map.
+	 *
+	 * @return array
+	 */
+	private function map_taxonomies_for_js( $taxonomy ) {
+		$primary_term = new WPSEO_Primary_Term( $taxonomy->name, get_the_ID() );
```

Comment:
If you made a wrapper for this, see comment on [line 81](https://github.com/Yoast/wordpress-seo/pull/2746/files#diff-93ebfff2ef428ae807ba3678c7200502R81). You don't have to do the same logic.